pkg/apis/pipeline/v1beta1/pipeline_validation_test.go

```diff
@@ -1039,10 +1039,395 @@ func TestValidatePipelineWorkspaces_Failure(t *testing.T) {
 	}}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			err := validatePipelineWorkspaces(tt.workspaces, tt.tasks)
+			err := validatePipelineWorkspaces(tt.workspaces, tt.tasks, []PipelineTask{})
 			if err == nil {
 				t.Error("Pipeline.validatePipelineWorkspaces() did not return error, wanted error")
 			}
 		})
 	}
 }
+
+func TestValidatePipelineWithFinalTasks(t *testing.T) {
+	tests := []struct {
+		name            string
+		p               *Pipeline
+		failureExpected bool
+	}{{
+		name: "valid pipeline with final tasks",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Tasks: []PipelineTask{{
+					Name:    "non-final-task",
+					TaskRef: &TaskRef{Name: "non-final-task"},
+				}},
+				Finally: []PipelineTask{{
+					Name:    "final-task-1",
+					TaskRef: &TaskRef{Name: "final-task"},
+				}, {
+					Name: "final-task-2",
+					TaskSpec: &TaskSpec{
+						Steps: []Step{{
+							Container: corev1.Container{Name: "foo", Image: "bar"},
+						}},
+					},
+				}},
+			},
+		},
+		failureExpected: false,
+	}, {
+		name: "valid pipeline with resource declarations and their valid usage",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Resources: []PipelineDeclaredResource{{
+					Name: "great-resource", Type: PipelineResourceTypeGit,
+				}, {
+					Name: "wonderful-resource", Type: PipelineResourceTypeImage,
+				}},
+				Tasks: []PipelineTask{{
+					Name:    "non-final-task",
+					TaskRef: &TaskRef{Name: "bar-task"},
+					Resources: &PipelineTaskResources{
+						Inputs: []PipelineTaskInputResource{{
+							Name: "some-workspace", Resource: "great-resource",
+						}},
+						Outputs: []PipelineTaskOutputResource{{
+							Name: "some-image", Resource: "wonderful-resource",
+						}},
+					},
+					Conditions: []PipelineTaskCondition{{
+						ConditionRef: "some-condition",
+						Resources: []PipelineTaskInputResource{{
+							Name: "some-workspace", Resource: "great-resource",
+						}},
+					}},
+				}},
+				Finally: []PipelineTask{{
+					Name:    "foo",
+					TaskRef: &TaskRef{Name: "foo-task"},
+					Resources: &PipelineTaskResources{
+						Inputs: []PipelineTaskInputResource{{
+							Name: "some-workspace", Resource: "great-resource",
+						}},
+						Outputs: []PipelineTaskOutputResource{{
+							Name: "some-image", Resource: "wonderful-resource",
+						}},
+					},
+				}},
+			},
+		},
+		failureExpected: false,
+	}, {
+		name: "invalid pipeline without any non-final task but at least one final task",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Tasks: nil,
+				Finally: []PipelineTask{{
+					Name:    "final-task",
+					TaskRef: &TaskRef{Name: "final-task"},
+				}},
+			},
+		},
+		failureExpected: true,
+	}, {
+		name: "invalid pipeline without any non-final task but at least one final task",
```

Comment:
i think this might be the same as the case before it?

Reply:
first test case has `Tasks: nil` and this test case has `Tasks: []PipelineTask{{}}`, updated the description